compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,926 @@
+    resultLattices.push_back(resultLattice);
+  }
+
+  auto changeFunc = [&](DistributionLayout *lattice, ChangeResult changed) {
+    this->propagateIfChanged(lattice, changed);
+  };
+
+  enforcementTransferFunction(op, operandLattices, resultLattices, changeFunc);
+}
+
+void EnforceLayout::visitRegionSuccessors(RegionBranchOpInterface branch,
+                                          RegionBranchPoint branchPoint,
+                                          MutableArrayRef<OpOperand> operands) {
+  SmallVector<RegionSuccessor> successors;
+  branch.getSuccessorRegions(branchPoint, successors);
+  for (RegionSuccessor successor : successors) {
+    ValueRange inputs = successor.getSuccessorInputs();
+
+    // Get vector layouts for forwarded operands.
+    SmallVector<DistributionLayout *> forwardedLattices;
+    SmallVector<OpOperand *> forwardedOperands;
+    for (OpOperand &use : operands) {
+      Value operand = use.get();
+      if (isa<VectorType>(operand.getType())) {
+        forwardedLattices.push_back(getLatticeElement(operand));
+        forwardedOperands.push_back(&use);
+      }
+    }
+
+    // Get vector layouts for input operands.
+    SmallVector<const DistributionLayout *> inputLattices;
+    for (Value operand : inputs) {
+      if (isa<VectorType>(operand.getType())) {
+        inputLattices.push_back(getLatticeElement(operand));
+      }
+    }
+
+    // Both should have same number of vector operands.
+    assert(forwardedLattices.size() == inputLattices.size() &&
+           "Number of forwarded operands and inputs should match");
+
+    // Propagate the layouts.
+    int64_t curr = 0;
+    for (auto [forwardedLattice, inputLattice] :
+         llvm::zip(forwardedLattices, inputLattices)) {
+      forwardedLattice->resolveWithPossibleConflict(inputLattice,
+                                                    *forwardedOperands[curr]);
+      curr++;
+    }
+  }
+}
+
+DistributionLayout *EnforceLayout::getLatticeElement(Value val) {
+  // Add dependency of operation on the analysis state.
+  assert(isa<VectorType>(val.getType()) && "Lattice value should be a vector");
+  DistributionLayout *layout =
+      DataFlowAnalysis::getOrCreate<DistributionLayout>(val);
+  // Subscribe this analysis to updates of the lattice.
+  layout->subscribeEnforcement(this);
+  return layout;
+}
+
+/// ==========================================================================
+///        VectorLayoutAnalysis
+/// ==========================================================================
+
+LogicalResult VectorLayoutAnalysis::run() {
+  // The order of loading matters here, because propagateLayout does anchoring
+  // initialization which needs the lattice to know both enforcement and
+  // propagation.
+  solver.load<EnforceLayout>(root->getContext());
+  solver.load<PropagateLayout>(anchors, root->getContext());
+  return solver.initializeAndRun(root);
+}
+
+VectorLayoutInterface VectorLayoutAnalysis::getLayout(Value val) {
```

Comment:
std::optional here as well

Reply:
VectorLayoutInterface is an attribute interface. Attributes already have a possible null value, so not sure if we need another optional on top of that.